### PR TITLE
configure: change install locations if P11 kit detected

### PR DESCRIPTION
**Pull request by williamcroberts**
Get working with P11 kit
Fixes: #28

### CODE REVIEW

**Comment by williamcroberts** [comment]
@dwmw2 could you help review?

File: configure.ac [hunk 1]
```diff
@@ -105,6 +105,39 @@ AS_IF([test "$unit" != "no"],
...
+
+  PKG_CHECK_VAR([P11_CONFIGS_PATH], [p11-kit-1], [p11_module_configs])
+  AC_MSG_CHECKING([p11 configs path])
+  AS_IF([test "x$P11_CONFIGS_PATH" = "x"], [
+    AC_MSG_FAILURE([Unable to identify p11 configs path.])
+  ])
+  AC_MSG_RESULT([$P11_CONFIGS_PATH])
+])
+
+# if P11 kit is installed set havep11kit=$enableval then:
```
Comment (williamcroberts) [inline configure.ac]: this comment is not correct

File: misc/p11-kit/tpm2-pkcs11 [hunk 2]
```diff
@@ -0,0 +1,3 @@
+module: tpm2-pkcs11.so
+
+critical: no
```
Comment (williamcroberts) [inline misc/p11-kit/tpm2-pkcs11]: should this have a newline?

**Comment by dwmw2** [comment]
Strictly speaking, RFC7512 URIs and p11-kit installation are orthogonal.
The p11-kit module file simply makes the module available; it doesn't *have* to be accessed via the URI. But that's good practice too, and is included in Fedora packaging guidelines too.

**Comment by dwmw2** [comment]
```
$ git pull github.com:williamcroberts/tpm2-pkcs11 issue-28-p11-kit
$ make
$ make install
$ p11-kit list-modules
```
Doesn't work. Looks like you've allowed it to prepend your own build's `$DESTDIR` to the location. Since all my TSS2 stuff is built with  `--prefix=/opt/tss2` I got `/opt/tss2/usr/share/p11-kit/modules/tpm2-pkcs11` created, which p11-kit won't find. It needs to be called `/usr/share/p11-kit/modules/tpm2-pkcs11.module` (note the `.module` suffix too).

That .module file then references the actual tpm2-pkcs11.so by name, which is the right thing to do because you might have both 32-bit and 64-bit versions in the system. So p11-kit (if it had found the .module file) would look in the appropriate 32-bit or 64-bit directory for it... and not find it there because you allowed that to get installed to $(DESTDIR) too :)

**Comment by dwmw2** [comment]
oh, and `strcmp("libtpm2-pkcs11.so", "tpm2-pkcs11.so")`

**Comment by dwmw2** [comment]
```
$ sudo mv /opt/tss2/usr/lib/x86_64-linux-gnu/pkcs11/libtpm2_pkcs11.so.0.0.0 /usr/lib/x86_64-linux-gnu/pkcs11/tpm2-pkcs11.so
$ sudo mv /opt/tss2/usr/share/p11-kit/modules/tpm2-pkcs11 /usr/share/p11-kit/modules/tpm2-pkcs11.module
$ $ p11-kit list-modules
ERROR: Could not stat db "tpm2_pkcs11.sqlite3" under store "/home/dwmw", error: No such file or directory
ERROR: Consider exporting TPM2_PKCS11_STORE to point to a valid store sirectory
ERROR: getting token count: SQL logic error
```

Yay, that looks better (look at using no-version to drop the .0.0.0 part of your soname, btw). Better to make it fail silently if the storage isn't there, too.

That's about as far as I've got with the token, because...

```
$ ./tpm2_ptool.py  init
Traceback (most recent call last):
  File "./tpm2_ptool.py", line 644, in __call__
    ctx = tpm2.createprimary(ownerauth, pobjauthhash['hash'])
  File "./tpm2_ptool.py", line 144, in createprimary
    raise RuntimeError("Could not execute tpm2_createprimary: %s" % stderr)
RuntimeError: Could not execute tpm2_createprimary: ERROR: Could not convert port to a 16 bit unsigned number, got: hex:d947ea342140421a86bb7ad6ced1fd176708a555c2960d6feac2b91620c76152
Failed to initialize tcti context: 0x1

Could not execute tpm2_createprimary: ERROR: Could not convert port to a 16 bit unsigned number, got: hex:d947ea342140421a86bb7ad6ced1fd176708a555c2960d6feac2b91620c76152
Failed to initialize tcti context: 0x1


```

**Comment by dwmw2** [comment]
Assuming you *do* have it working, you ought to now be able to use `p11tool --list-tokens` and find specific keys in your token by following instructions like the ones at http://www.infradead.org/openconnect/pkcs11.html

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: David Woodhouse [mailto:notifications@github.com]
> Sent: Tuesday, October 9, 2018 3:58 PM
> To: tpm2-software/tpm2-pkcs11 <tpm2-pkcs11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Author
> <author@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-pkcs11] [WIP] configure: change install
> locations if P11 kit detected (#30)
> 
> $ sudo mv /opt/tss2/usr/lib/x86_64-linux-gnu/pkcs11/libtpm2_pkcs11.so.0.0.0
> /usr/lib/x86_64-linux-gnu/pkcs11/tpm2-pkcs11.so
> $ sudo mv /opt/tss2/usr/share/p11-kit/modules/tpm2-pkcs11 /usr/share/p11-
> kit/modules/tpm2-pkcs11.module
> $ $ p11-kit list-modules
> ERROR: Could not stat db "tpm2_pkcs11.sqlite3" under store "/home/dwmw",
> error: No such file or directory
> ERROR: Consider exporting TPM2_PKCS11_STORE to point to a valid store
> sirectory

Spelling and should have it fail there, not sure why the error below is still propgating.

> ERROR: getting token count: SQL logic error
> 
> 
> Yay, that looks better (look at using no-version to drop the .0.0.0 part of your
> soname, btw). Better to make it fail silently if the storage isn't there, too.

I'll have to look into the soname part.... my autoconf/make wizardry is lacking

As far as silent failure on "storage isn't there".

I guess just return that it hasn't been initialized via
CKF_TOKEN_INITIALIZED and fail in all API calls?

As far as the error message goes, I could move it to a LOGV call which hides it
By default.

> 
> That's about as far as I've got with the token, because...
> 
> $ ./tpm2_ptool.py  init
> Traceback (most recent call last):
>   File "./tpm2_ptool.py", line 644, in __call__
>     ctx = tpm2.createprimary(ownerauth, pobjauthhash['hash'])
>   File "./tpm2_ptool.py", line 144, in createprimary
>     raise RuntimeError("Could not execute tpm2_createprimary: %s" % stderr)
> RuntimeError: Could not execute tpm2_createprimary: ERROR: Could not convert
> port to a 16 bit unsigned number, got:
> hex:d947ea342140421a86bb7ad6ced1fd176708a555c2960d6feac2b91620c76152
> Failed to initialize tcti context: 0x1
> 
> Could not execute tpm2_createprimary: ERROR: Could not convert port to a 16 bit
> unsigned number, got:
> hex:d947ea342140421a86bb7ad6ced1fd176708a555c2960d6feac2b91620c76152
> Failed to initialize tcti context: 0x1

I am going to guess you're not on the tip-of-master for tpm2-tools.

**Comment by dwmw2** [comment]
> I am going to guess you're not on the tip-of-master for tpm2-tools.

True. 
```
 $ ./tpm2_ptool.py init
Created a primary object of id: 1
 $ p11tool --list-tokens
ERROR: Could not stat db "tpm2_pkcs11.sqlite3" under store "/home/dwmw", error: No such file or directory
ERROR: Consider exporting TPM2_PKCS11_STORE to point to a valid store sirectory
 $ export TPM2_PKCS11_STORE=~/.tpm2_pkcs11
 $ p11tool --list-tokens
Segmentation fault
```
```
(gdb) bt
#0  default_mutex_create (mutex=0x0) at src/lib/mutex.c:61
#1  0x00007ffff5842171 in session_table_new (t=t@entry=0x7ffff5a4d298 <global>) at src/lib/session_table.c:29
#2  0x00007ffff5840ea4 in session_init () at src/lib/session.c:28
#3  0x00007ffff583fb6f in general_init (init_args=init_args@entry=0x5555557b3c00) at src/lib/general.c:217
#4  0x00007ffff5839934 in C_Initialize (init_args=0x5555557b3c00) at src/pkcs11.c:40
#5  0x00007ffff6d3a7c5 in ?? () from /usr/lib/x86_64-linux-gnu/libp11-kit.so.0
#6  0x00007ffff6d3a983 in ?? () from /usr/lib/x86_64-linux-gnu/libp11-kit.so.0
#7  0x00007ffff6d543a0 in ?? () from /usr/lib/x86_64-linux-gnu/libp11-kit.so.0
#8  0x00007ffff5e76b4f in ffi_closure_unix64_inner () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#9  0x00007ffff5e76f16 in ffi_closure_unix64 () from /usr/lib/x86_64-linux-gnu/libffi.so.6
#10 0x00007ffff6d3d2c3 in p11_kit_modules_initialize () from /usr/lib/x86_64-linux-gnu/libp11-kit.so.0
#11 0x00007ffff6d3d5b1 in p11_kit_modules_load_and_initialize () from /usr/lib/x86_64-linux-gnu/libp11-kit.so.0
#12 0x00007ffff7ad422a in ?? () from /usr/lib/x86_64-linux-gnu/libgnutls.so.30
#13 0x00007ffff7ad5b17 in gnutls_pkcs11_init () from /usr/lib/x86_64-linux-gnu/libgnutls.so.30
#14 0x000055555555f581 in ?? ()
#15 0x00007ffff726cb97 in __libc_start_main (main=0x55555555f4e0, argc=2, argv=0x7fffffffdca8, init=<optimised out>, fini=<optimised out>, rtld_fini=<optimised out>, 
    stack_end=0x7fffffffdc98) at ../csu/libc-start.c:310
#16 0x000055555555fe6a in ?? ()
```
But now I suppose this should be a different issue.

**Comment by dwmw2** [comment]
You're not using your own locking primitives, are you? You have to use the ones passed in by the caller.

**Comment by dwmw2** [comment]
I managed to work around the crash with this:
```
--- a/src/lib/session_table.c
+++ b/src/lib/session_table.c
@@ -26,7 +26,7 @@ CK_RV session_table_new(session_table **t) {
         return CKR_HOST_MEMORY;
     }
 
-    CK_RV rv = mutex_create(x->lock);
+    CK_RV rv = mutex_create(&x->lock);
     if (rv != CKR_OK) {
         session_table_free(x);
         return rv;
```

I still wasn't able to add a token though. And then calliing `tpm2_ptool.py init` to start from scratch seemed to run out of NV storage. Is it really using NV storage for anything? Why? Can't it save everything in the file? After resetting the TPM completely and really starting from scratch, I still can't add a token:
```
dwmw@u3832b3a9db3152:~/git/tpm2-pkcs11/tools$ ./tpm2_ptool.py init --pobj-pin=myobjpin
Created a primary object of id: 1
dwmw@u3832b3a9db3152:~/git/tpm2-pkcs11/tools$ ./tpm2_ptool.py addtoken --pid=1 --pobj-pin=mypobjpin --sopin=mysopin --userpin=myuserpin --label=label
Traceback (most recent call last):
  File "./tpm2_ptool.py", line 1277, in <module>
    main()
  File "./tpm2_ptool.py", line 1274, in main
    commandlet.get()[which](d)
  File "./tpm2_ptool.py", line 796, in __call__
    pobjauthhash = c.decrypt(pobject['pobjauth'])
  File "./tpm2_ptool.py", line 123, in decrypt
    plaintext = decryptor.update(ciphertext) + decryptor.finalize()
  File "/usr/lib/python2.7/dist-packages/cryptography/hazmat/primitives/ciphers/base.py", line 198, in finalize
    data = self._ctx.finalize()
  File "/usr/lib/python2.7/dist-packages/cryptography/hazmat/backends/openssl/ciphers.py", line 164, in finalize
    raise InvalidTag
cryptography.exceptions.InvalidTag

```

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: David Woodhouse [mailto:notifications@github.com]
> Sent: Tuesday, October 9, 2018 6:28 PM
> To: tpm2-software/tpm2-pkcs11 <tpm2-pkcs11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Author
> <author@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-pkcs11] [WIP] configure: change install
> locations if P11 kit detected (#30)
> 
> I managed to work around the crash with this:
> 
> --- a/src/lib/session_table.c
> +++ b/src/lib/session_table.c
> @@ -26,7 +26,7 @@ CK_RV session_table_new(session_table **t) {
>          return CKR_HOST_MEMORY;
>      }
> 
> -    CK_RV rv = mutex_create(x->lock);
> +    CK_RV rv = mutex_create(&x->lock);
>      if (rv != CKR_OK) {
>          session_table_free(x);
>          return rv;
> 
> 
> I still wasn't able to add a token though. And then calliing tpm2_ptool.py init to
> start from scratch seemed to run out of NV storage. Is it really using NV storage
> for anything? Why? 

To use the owner hierarchy of the TPM requires the owner hierarchy password, which
(from what I have been told) is typically tightly controlled by something like an IT
organization. Thus to create a primary object in the hierarchy would require this on
every invocation, which is unrealistic. To work around this design, tpm2-pkcs11 creates
1 **persistent** primary object under the owner hierarchy. 

The design allows for the ability In the future, if a persistent primary
object already exists, the user can just tell us what to use and
we won't create a new one.

> Can't it save everything in the file? After resetting the TPM

Everything else is in a file in the store.

> completely and really starting from scratch, I still can't add a token:
> 
> dwmw@u3832b3a9db3152:~/git/tpm2-pkcs11/tools$ ./tpm2_ptool.py init --pobj-
> pin=myobjpin Created a primary object of id: 1
> dwmw@u3832b3a9db3152:~/git/tpm2-pkcs11/tools$ ./tpm2_ptool.py addtoken
> --pid=1 --pobj-pin=mypobjpin --sopin=mysopin --userpin=myuserpin --
> label=label Traceback (most recent call last):
>   File "./tpm2_ptool.py", line 1277, in <module>
>     main()
>   File "./tpm2_ptool.py", line 1274, in main
>     commandlet.get()[which](d)
>   File "./tpm2_ptool.py", line 796, in __call__
>     pobjauthhash = c.decrypt(pobject['pobjauth'])
>   File "./tpm2_ptool.py", line 123, in decrypt
>     plaintext = decryptor.update(ciphertext) + decryptor.finalize()
>   File "/usr/lib/python2.7/dist-
> packages/cryptography/hazmat/primitives/ciphers/base.py", line 198, in finalize
>     data = self._ctx.finalize()
>   File "/usr/lib/python2.7/dist-
> packages/cryptography/hazmat/backends/openssl/ciphers.py", line 164, in
> finalize
>     raise InvalidTag
> cryptography.exceptions.InvalidTag
> 

Oh it's obvious what the problem is (no its not, need better error message). You're using
The wrong pob-pin between commandlets init and addtoken. They need to be the same.

**Comment by dwmw2** [comment]
Hm, the tpm2-tss-engine does rely on creating a new primary every single time, for use with keys stored in PEM. 

>  You're using The wrong pob-pin between commandlets init and addtoken. They need to be the same.

Oops, sorry. Fixed now, and it worked. Except p11tool --list-tokens never seems to return...
```
 $ ./tpm2_ptool.py addtoken --pid=1 --pobj-pin=myobjpin --sopin=mysopin --userpin=myuserpin --label=label
auto-detecting TPM encryptdecrypt interface for wrapping key usage
Using "Software" based object authorization protections
Created token: label
 $ p11tool --list-tokens
Token 0:
	URL: pkcs11:model=p11-kit-trust;manufacturer=PKCS%2311%20Kit;serial=1;token=System%20Trust
	Label: System Trust
	Type: Trust module
	Manufacturer: PKCS#11 Kit
	Model: p11-kit-trust
	Serial: 1
	Module: p11-kit-trust.so

```

Looks like it deadlocked:

```
#0  __lll_lock_wait () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:135
#1  0x00007f9cd131e023 in __GI___pthread_mutex_lock (mutex=0x55ca0a6b4f50) at ../nptl/pthread_mutex_lock.c:78
#2  0x00007f9cd1106019 in default_mutex_lock (mutex=<optimised out>) at src/lib/mutex.c:81
#3  0x00007f9cd11081e0 in mutex_lock_fatal (mutex=<optimised out>) at src/lib/mutex.h:67
#4  session_table_lock (t=t@entry=0x55ca0a6b52a0) at src/lib/session_table.c:53
#5  0x00007f9cd11082a3 in session_table_get_cnt (t=0x55ca0a6b52a0, is_rw=is_rw@entry=false) at src/lib/session_table.c:61
#6  0x00007f9cd1106fb0 in check_max_sessions (is_rw=<optimised out>) at src/lib/session.c:38
#7  session_open (slot_id=slot_id@entry=1, flags=flags@entry=4, application=application@entry=0x0, notify=notify@entry=0x0, session=session@entry=0x7fff70737070) at src/lib/session.c:70
#8  0x00007f9cd1100283 in C_OpenSession (slotID=1, flags=4, application=0x0, notify=0x0, session=0x7fff70737070) at src/pkcs11.c:105

```

**Comment by williamcroberts** [comment]
Yeah the threading stuff needs to be sorted. Ill spend some time with this tomorrow.

In the meantime, some fixes ready for master that have come out of this discussion can be found at:
https://github.com/tpm2-software/tpm2-pkcs11/pull/31

Thank you for kicking the tires on this, this is great feedback. Much appreciated.

**Comment by dwmw2** [comment]
NP. I'm spending tomorrow in a tin can so won't be very responsive. I think you're most of the way to being able to use this with p11tool. If you look at the OpenConnect tests in http://git.infradead.org/users/dwmw2/openconnect.git/blob/tpm2:/tests/Makefile.am you'll see three different ways it imports keys into a token (in that case a SoftHSM token). Then it uses them from OpenConnect. It should all translate relatively straightforwardly to doing it on your own token instead of SoftHSM.

**Comment by jejb** [comment]
On Wed, 2018-10-10 at 13:29 +0000, David Woodhouse wrote:
> @AndreasFuchsSIT @jejb see discussion of feasibility of creating new
> primaries on demand, in practice...

If I interpret correctly: the problem you foresee is that
TPM2_CreatePrimary can only be run for the SPS if you know the owner
authorization and you might be operating in environments where you
don't?

The solution for that is to use the RSA well known storage primary at
81000001 (as the PC provisioning guide states)

https://trustedcomputinggroup.org/resource/tcg-tpm-v2-0-provisioning-gu
idance/

The guidance requires this to have an empty password even when the
owner password isn't.  So the solution I came up with is simply to
allow the key creator to give a parent and assume I can run
TPM2_CreatePrimary with an empty password if none is specified and
error out if that fails.

James

**Comment by idesai** [comment]
+1 to what @James said and it is in line with the current implementation, just need that persistenthandle to be 0x81000001. Need to include in the documentation about the TCG provisioning guidance.
If there already exists a persistent handle with a key authentication set, we could inform the user of the TCG provisioning guidance and still give user the choice to use next available handle since this is not really a mandate.
@James, is there any guidance on what createprimary object’s key attributes should be like?

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Imran Desai [mailto:notifications@github.com]
> Sent: Wednesday, October 10, 2018 8:20 AM
> To: tpm2-software/tpm2-pkcs11 <tpm2-pkcs11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Author
> <author@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-pkcs11] [WIP] configure: change install
> locations if P11 kit detected (#30)
> 
> +1 to what @James said and it is in line with the current implementation, just
> need that persistenthandle to be 0x81000001. Need to include in the
> documentation about the TCG provisioning guidance.
> If there already exists a persistent handle with a key authentication set, we could
> inform the user of the TCG provisioning guidance and still give user the choice to
> use next available handle since this is not really a mandate.
> @James, is there any guidance on what createprimary object’s key attributes
> should be like?
> 
> From: James Bottomley <notifications@github.com>
> Reply-To: tpm2-software/tpm2-pkcs11 <reply@reply.github.com>
> Date: Wednesday, October 10, 2018 at 7:06 AM
> To: tpm2-software/tpm2-pkcs11 <tpm2-pkcs11@noreply.github.com>
> Cc: Subscribed <subscribed@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-pkcs11] [WIP] configure: change install
> locations if P11 kit detected (#30)
> 
> On Wed, 2018-10-10 at 13:29 +0000, David Woodhouse wrote:
> > @AndreasFuchsSIT @jejb see discussion of feasibility of creating new
> > primaries on demand, in practice...
> 
> If I interpret correctly: the problem you foresee is that TPM2_CreatePrimary can
> only be run for the SPS if you know the owner authorization and you might be
> operating in environments where you don't?
> 
> The solution for that is to use the RSA well known storage primary at
> 81000001 (as the PC provisioning guide states)

Yes this type of scenario is forseen in the design, but currently not implemented.
The "tpm2_ptool.py init" command can just take in the proper handle and auth value
to include the empty auth and populate the db.

> 
> https://trustedcomputinggroup.org/resource/tcg-tpm-v2-0-provisioning-gu
> idance/
> 
> The guidance requires this to have an empty password even when the owner
> password isn't. So the solution I came up with is simply to allow the key creator to
> give a parent and assume I can run TPM2_CreatePrimary with an empty

createprimary with an ampty password, did you mean create here?

**Comment by jejb** [comment]
On Wed, 2018-10-10 at 08:19 -0700, Imran Desai wrote:
> @James, is there any guidance on what createprimary object’s key
> attributes should be like?

Sort of, it's section 7.5.1 Storage Primary Key (SRK) Templates

But, like most TCG documents it doesn't give full information, you have
to refer back to the EK Credential Profile 2.0

https://trustedcomputinggroup.org/resource/tcg-ek-credential-profile-for-tpm-family-2-0/

Don't rely on the links long term because the TCG reorganizes its
website every few months.

James

**Comment by jejb** [comment]
On Wed, 2018-10-10 at 08:47 -0700, William Roberts wrote:
> > The guidance requires this to have an empty password even when the
> > owner password isn't. So the solution I came up with is simply to
> > allow the key creator to give a parent and assume I can run
> > TPM2_CreatePrimary with an empty
> 
> createprimary with an ampty password, did you mean create here?

No, it's a key derived from a seed not a key the TPM creates randomly.

James

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: James Bottomley [mailto:notifications@github.com]
> Sent: Wednesday, October 10, 2018 9:42 AM
> To: tpm2-software/tpm2-pkcs11 <tpm2-pkcs11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Author
> <author@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-pkcs11] [WIP] configure: change install
> locations if P11 kit detected (#30)
> 
> On Wed, 2018-10-10 at 08:47 -0700, William Roberts wrote:
> > > The guidance requires this to have an empty password even when the
> > > owner password isn't. So the solution I came up with is simply to
> > > allow the key creator to give a parent and assume I can run
> > > TPM2_CreatePrimary with an empty
> >
> > createprimary with an ampty password, did you mean create here?
> 
> No, it's a key derived from a seed not a key the TPM creates randomly.

Above you state to the affect you allow the key creator to give a parent and assume I
can run tpm2_createprimary with an empty authorization. Initially I was assuming the parent
you were referring to was the RSA key at 0x81000001 as stated in the provisioning guidance.

However it seems based on your response the parent is one of the hierarchies and your hoping
It's an empty auth.

This is where the flexibility of the design shines. None of the scenarios mentioned are
unsupportable.

**Comment by jejb** [comment]
On Wed, 2018-10-10 at 16:52 +0000, William Roberts wrote:
> > > createprimary with an ampty password, did you mean create here?
> > 
> > No, it's a key derived from a seed not a key the TPM creates
> > randomly.
> 
> Above you state to the affect you allow the key creator to give a
> parent and assume I can run tpm2_createprimary with an empty
> authorization.

No, you have to run createprimary on the SPS with the owner auth.  If
that's empty, then you may use empty auth.

The conventional provisioning for a system where the owner auth is non
empty is to have the RSA primary provisioned at 81000001 with empty
auth.

>  Initially I was assuming the parent you were referring to was the
> RSA key at 0x81000001 as stated in the provisioning guidance.
> 
> However it seems based on your response the parent is one of the
> hierarchies and your hoping It's an empty auth.

The parent can be any key in the storage hierarchy ... it's not limited
at all.  However, if no parent at all is specified, the tpm engine will
simply run TPM2_createprimary on the SPS with empty auth.  If a parent
is specified, it will use it.  Conventionally if an owner password is
set the user specifies 81000001 as the parent which means we use the
RSA derived key of the SPS.

James

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: James Bottomley [mailto:notifications@github.com]
> Sent: Wednesday, October 10, 2018 10:05 AM
> To: tpm2-software/tpm2-pkcs11 <tpm2-pkcs11@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Author
> <author@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-pkcs11] [WIP] configure: change install
> locations if P11 kit detected (#30)
> 
> On Wed, 2018-10-10 at 16:52 +0000, William Roberts wrote:
> > > > createprimary with an ampty password, did you mean create here?
> > >
> > > No, it's a key derived from a seed not a key the TPM creates
> > > randomly.
> >
> > Above you state to the affect you allow the key creator to give a
> > parent and assume I can run tpm2_createprimary with an empty
> > authorization.
> 
> No, you have to run createprimary on the SPS with the owner auth. If that's
> empty, then you may use empty auth.
> 
> The conventional provisioning for a system where the owner auth is non empty is
> to have the RSA primary provisioned at 81000001 with empty auth.
> 
> > Initially I was assuming the parent you were referring to was the RSA
> > key at 0x81000001 as stated in the provisioning guidance.
> >
> > However it seems based on your response the parent is one of the
> > hierarchies and your hoping It's an empty auth.
> 
> The parent can be any key in the storage hierarchy ... it's not limited at all.
> However, if no parent at all is specified, the tpm engine will simply run
> TPM2_createprimary on the SPS with empty auth. If a parent is specified, it will
> use it. Conventionally if an owner password is set the user specifies 81000001 as
> the parent which means we use the RSA derived key of the SPS.

Yes were on the same page. The confusion was between
the two cases of an empty auth, one being on a hierarchy and the other
being on a primary object and how you handle them in the engine. Your
segue between the two cases wasn't particularly clear to me.

All these scenarios are trivial to support in the tpm2_ptool.py init command.

**Comment by williamcroberts** [comment]
@jejb I made an issue tracker to capture this work item so it doesn't disappear on us: https://github.com/tpm2-software/tpm2-pkcs11/issues/35

**Pull request merged**